texar/modules/connectors/connectors.py

```diff
@@ -0,0 +1,786 @@
+            or a ``tuple`` of ``torch.Size``.
+            For example, to transform inputs to have decoder state size, set
+            :python:`output_size=decoder.state_size`.
+            If :attr:`output_size` is ``tuple`` ```(1, 2, 3)```, then the
+            output structure will be
+            ``([batch_size * 1], [batch_size * 2], [batch_size * 3])``.
+            If :attr:`output_size` is ``torch.Size([1, 2, 3])``, then the
+            output structure will be ``[batch_size, 1, 2, 3]``.
+        hparams (dict, optional): Hyperparameters. Missing
+            hyperparamerter will be set to default values. See
+            :meth:`default_hparams` for the hyperparameter sturcture and
+            default values.
+    This connector does not have trainable parameters.
+
+    Example:
+
+        .. code-block:: python
+
+            state_size = (1, 2, 3)
+            connector = ConstantConnector(state_size, hparams={"value": 1.})
+            one_state = connector(batch_size=64)
+            # `one_state` structure: (Tensor_1, Tensor_2, Tensor_3),
+            # Tensor_1.size() == torch.Size([64, 1])
+            # Tensor_2.size() == torch.Size([64, 2])
+            # Tensor_3.size() == torch.Size([64, 3])
+            # Tensors are filled with 1.0.
+            size = torch.Size([1, 2, 3])
+            connector_size = ConstantConnector(size, hparams={"value": 2.})
+            size_state = connector_size_0(batch_size=64)
+            # `size_state` structure: Tensor with size [64, 1, 2, 3].
+            # Tensor is filled with 2.0.
+
+    """
+    def __init__(self,
+                 output_size: OutputSize,
+                 hparams: HParamsType = None):
+
+        ConnectorBase.__init__(self, output_size, hparams)
+
+    @staticmethod
+    def default_hparams() -> dict:
+        r"""Returns a dictionary of hyperparameters with default values.
+
+        .. code-block:: python
+
+            {
+                "value": 0.,
+                "name": "constant_connector"
+            }
+
+        Here:
+
+        "value" : float
+            The constant scalar that the output tensor(s) has.
+        "name" : str
+            Name of the connector.
+        """
+        return {
+            "value": 0.,
+            "name": "constant_connector"
+        }
+
+    def forward(self, # type: ignore
+                batch_size: Union[int, torch.Tensor]) -> Any:
+        """Creates output tensor(s) that has the given value.
+
+        Args:
+            batch_size: An ``int`` or ``int`` scalar ``Tensor``, the
+                batch size.
+            value (optional): A scalar, the value that the output tensor(s)
+                has. If ``None``, ``"value"`` in :attr:`hparams` is used.
+
+        Returns:
+            A (structure of) ``Tensor`` whose structure is the same as
+            :attr:`output_size`, with value speicified by
+            ``value`` or :attr:`hparams`.
+        """
+        value_ = self.hparams.value
```

Comment:
The reason we removed `value` from the argument is because we can infer the module device from the value in HParams. This requires us to convert value to a tensor in `__init__`, and registering it as a buffer of the module.

Reply:
PS. A buffer of a `nn.Module` is a tensor that is part of the module state but not a parameter. Buffers are moved along with parameters with `.cuda()` or `.to(device=)` is called. Please see <https://pytorch.org/docs/stable/nn.html#torch.nn.Module.register_buffer>

Reply:
Thanks! I know how to it next time if we want to infer the device etc.